Add async_function to TableFunction

carlopi opened the pull request:
This is scaffolding towards allowing BLOCKING during scans, but I think given the surface to separate this in different PRs.

Core of this PR is allowing an additional TableFunction callback, `asycn_function`, that takes precedence over regular `function` and already bakes in (allowing in the future more customisations) the `SourceResultType` that should be returned.

MultiFileScans are already moved to new signature, but currently all functions are expected to have an implementation equivalent to:
```
return chunk.size() ? SourceResultType::HAVE_MORE_OUTPUT : SourceResultType::FINISHED;
```
and returning BLOCKING is at the moment invalid (and checked).

This PR should not change any code actual behaviour.

carlopi commented on extension/parquet/parquet_reader.cpp:
```diff
@@ -1256,13 +1256,24 @@ void ParquetReader::InitializeScan(ClientContext &context, ParquetReaderScanStat
+		bool keep_going;
+		SourceResultType res = ScanInternal(context, state, result, interrupt_state, keep_going);
+
+		if (res == SourceResultType::BLOCKED)
```
thanks, unsure of why this passed code quality though: https://github.com/duckdb/duckdb/actions/runs/18344716790

I sort of assume formatter is done automagically.

carlopi commented on extension/parquet/parquet_reader.cpp:
```diff
@@ -1370,15 +1384,18 @@ bool ParquetReader::ScanInternal(ClientContext &context, ParquetReaderScanState
-		return false; // end of last group, we are done
+		// end of last group, we are done
+		keep_going = false;
+		return SourceResultType::HAVE_MORE_OUTPUT;
```
I'll double check this, thanks

Tishj commented on src/include/duckdb/function/table_function.hpp:
```diff
@@ -280,6 +281,8 @@ typedef unique_ptr<LocalTableFunctionState> (*table_function_init_local_t)(Execu
 typedef unique_ptr<BaseStatistics> (*table_statistics_t)(ClientContext &context, const FunctionData *bind_data,
                                                          column_t column_index);
 typedef void (*table_function_t)(ClientContext &context, TableFunctionInput &data, DataChunk &output);
+typedef SourceResultType (*async_table_function_t)(ClientContext &context, TableFunctionInput &data, DataChunk &output,
```
As I've mentioned, I don't think this should directly take the `InterruptState`, performing the block should instead be a separate callback.

This is because there are scenarios where an operator is not allowed to block.
I think the logic here should be: return `SourceResultType::BLOCKED` to express intent to block, then the executor/global state of the operator is used to check whether the block is allowed to happen, when the block is allowed to happen, the callback to perform the block is issued, which receives the `InterruptState`

carlopi committed:
Handle PR review

carlopi committed:
More strict format

carlopi committed:
Add async_function, that takes precedence and allows more control over behaviors

This is scaffolding towards allowing BLOCKING during scans, but I think given the surface to separate
this in different PRs.

MultiFileScans are already moved to new signature, but currently all functions are expected to have an implementation equivalent to:
```
return chunk.size() ? SourceResultType::HAVE_MORE_OUTPUT : SourceResultType::FINISHED;
```
and returing BLOCKING is at the moment invalid (and checked).

This should not change any code actual behaviour.

carlopi committed:
Fixup Scan in extension/parquet

Tishj commented:
Superseded by https://github.com/duckdb/duckdb/pull/19422
